// frontend/src/pages/admin/Community.tsx
import React, { useState } from 'react';
import { Trash2, ImageIcon, Video, FileText, Youtube, Link as LinkIcon } from 'lucide-react';
import { useAuth } from '../../contexts/AuthContext';
import axios from 'axios';

// Initialize Axios instance using your existing pattern
const apiaxis = axios.create({
  baseURL: import.meta.env.VITE_API_BASE_URL || 'http://localhost:5000',
});

// --- Types ---
type MediaType = 'image' | 'video' | 'gif' | 'pdf' | 'youtube' | 'link';

interface MediaItem {
  type: MediaType;
  url: string;
  thumbnail: string | null;
  meta?: {
    size?: string;
    duration?: string;
    format?: string;
  };
}

interface FormData {
  school_name: string;
  area: string;
  state: string;
  date: string;
  session: string;
  title: string;
  description: string;
}

const INDIAN_STATES = ["Telangana", "Andhra Pradesh", "Karnataka", "Maharashtra", "Delhi", "Tamil Nadu", "Other"];
const SESSIONS = ["Morning", "Afternoon", "Full Day", "Workshop", "Seminar"];

// --- Mock Upload Service (Replace in production with Cloudinary/S3) ---
const uploadToExternalService = async (
  file: File,
  token: string // ✅ add token here
): Promise<{ url: string; type: MediaType }> => {
  const formData = new FormData();
  formData.append('file', file);

  const response = await apiaxis.post('/api/upload/media', formData, {
    headers: {
      'Authorization': `Bearer ${token}`, // ✅ now accessible
    },
  });

  if (!response.data.success) {
    throw new Error(response.data.error || 'Upload failed');
  }

  const typeMap: Record<string, MediaType> = {
    image: 'image',
    video: 'video',
    pdf: 'pdf',
  };

  return {
    url: response.data.url,
    type: typeMap[response.data.type] || 'link',
  };
};

export default function UniversalContentForm() {
  const { token } = useAuth(); // ✅ Your existing context

  const [form, setForm] = useState<FormData>({
    school_name: '',
    area: '',
    state: '',
    date: '',
    session: 'Morning',
    title: '',
    description: '',
  });

  const [mediaItems, setMediaItems] = useState<MediaItem[]>([]);
  const [loading, setLoading] = useState(false);
  const [activeTab, setActiveTab] = useState<'upload' | 'link'>('upload');
  const [linkInput, setLinkInput] = useState('');

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => {
    const { name, value } = e.target;
    setForm(prev => ({ ...prev, [name]: value }));
  };

  const handleFileUpload = async (e: React.ChangeEvent<HTMLInputElement>) => {
  if (!e.target.files?.length) return;

   // ✅ Guard: ensure token exists
  if (!token) {
    alert('You must be logged in to upload files.');
    return;
  }

  const file = e.target.files[0];
  setLoading(true);

  try {
    // ✅ Pass token here
    const { url, type: detectedType } = await uploadToExternalService(file, token);
    
    const newItem: MediaItem = {
      type: detectedType,
      url,
      thumbnail: detectedType === 'image' ? url : null,
      meta: {
        size: `${(file.size / 1024 / 1024).toFixed(2)} MB`,
        format: file.type,
      },
    };
    setMediaItems(prev => [...prev, newItem]);
  } catch (error) {
    console.error('Upload failed:', error);
    alert('Upload failed. Please try again.');
  } finally {
    setLoading(false);
    e.target.value = '';
  }
};

  const handleAddLink = (type: 'youtube' | 'link') => {
  const cleanInput = linkInput.trim();
  if (!cleanInput) {
    alert('Please enter a valid URL');
    return;
  }

  let thumbnail = null;
  if (type === 'youtube') {
    const match = cleanInput.match(/[?&]v=([^&#]*)/);
    const videoId = match?.[1];
    if (videoId) {
      // ✅ Remove space!
      thumbnail = `https://img.youtube.com/vi/${videoId}/0.jpg`;
    }
    // Even if no videoId, store the URL!
  }

  const newItem: MediaItem = { 
    type, 
    url: cleanInput, // ✅ Always store the input
    thumbnail 
  };
  setMediaItems(prev => [...prev, newItem]);
  setLinkInput('');
};

  const removeMedia = (index: number) => {
    setMediaItems(prev => prev.filter((_, i) => i !== index));
  };

  const handleSubmit = async (e: React.FormEvent) => {
    e.preventDefault();
    if (loading) return;

    // Validation
    if (!form.school_name.trim() || !form.title.trim()) {
      alert('School Name and Title are required.');
      return;
    }

    // Auth check (should not happen if route is protected, but safe)
    if (!token) {
      alert('You must be logged in to submit content.');
      return;
    }

    const payload = {
      school_name: form.school_name,
      area: form.area,
      state: form.state,
      date: form.date || null,
      session: form.session,
      title: form.title,
      description: form.description,
      media: mediaItems,
    };
    console.log("📤 Sending payload:", payload);

    try {
      setLoading(true);

      // ✅ Send to your backend with auth header
      const response = await apiaxis.post('/api/community', payload, {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
      });

      if (response.status === 201) {
        alert('Content submitted successfully!');
        // Reset form
        setForm({
          school_name: '',
          area: '',
          state: '',
          date: '',
          session: 'Morning',
          title: '',
          description: '',
        });
        setMediaItems([]);
      }
    } catch (error: unknown) {
      console.error('Submission error:', error);
      const message = axios.isAxiosError(error)
        ? error.response?.data?.error || error.message || 'Failed to submit content. Please try again.'
        : 'Failed to submit content. Please try again.';
      alert(message);
    } finally {
      setLoading(false);
    }
  };

  return (
    <div className="max-w-4xl mx-auto p-6 bg-white rounded-xl shadow-md border border-gray-200 my-8">
      <h2 className="text-2xl font-bold text-gray-800 mb-6">Add Client Services Data</h2>

      <form onSubmit={handleSubmit} className="space-y-6">
        {/* Row 1: School Info */}
        <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
          <div>
            <label className="block text-sm font-medium text-gray-700 mb-1">School Name *</label>
            <input
              required
              name="school_name"
              value={form.school_name}
              onChange={handleInputChange}
              className="w-full p-2 border rounded-md focus:ring-2 focus:ring-blue-500 outline-none"
              placeholder="e.g. Greenwood High"
            />
          </div>
          <div>
            <label className="block text-sm font-medium text-gray-700 mb-1">Area / Locality</label>
            <input
              name="area"
              value={form.area}
              onChange={handleInputChange}
              className="w-full p-2 border rounded-md"
              placeholder="e.g. Banjara Hills"
            />
          </div>
        </div>

        {/* Row 2: Details */}
        <div className="grid grid-cols-1 md:grid-cols-3 gap-4">
          <div>
            <label className="block text-sm font-medium text-gray-700 mb-1">State</label>
            <select
              name="state"
              value={form.state}
              onChange={handleInputChange}
              className="w-full p-2 border rounded-md bg-white"
            >
              <option value="">Select State</option>
              {INDIAN_STATES.map((s) => (
                <option key={s} value={s}>
                  {s}
                </option>
              ))}
            </select>
          </div>
          <div>
            <label className="block text-sm font-medium text-gray-700 mb-1">Date</label>
            <input
              type="date"
              name="date"
              value={form.date}
              onChange={handleInputChange}
              className="w-full p-2 border rounded-md"
            />
          </div>
          <div>
            <label className="block text-sm font-medium text-gray-700 mb-1">Session</label>
            <select
              name="session"
              value={form.session}
              onChange={handleInputChange}
              className="w-full p-2 border rounded-md bg-white"
            >
              {SESSIONS.map((s) => (
                <option key={s} value={s}>
                  {s}
                </option>
              ))}
            </select>
          </div>
        </div>

        {/* Title & Description */}
        <div>
          <label className="block text-sm font-medium text-gray-700 mb-1">Title *</label>
          <input
            name="title"
            value={form.title}
            onChange={handleInputChange}
            className="w-full p-2 border rounded-md"
          />
        </div>
        <div>
          <label className="block text-sm font-medium text-gray-700 mb-1">Description</label>
          <textarea
            name="description"
            value={form.description}
            onChange={handleInputChange}
            rows={3}
            className="w-full p-2 border rounded-md"
          />
        </div>

        <hr className="my-4" />

        {/* Media Gallery */}
        <div>
          <label className="block text-sm font-bold text-gray-700 mb-2">
            Media Gallery (Images, Videos, PDFs, Links)
          </label>

          <div className="bg-gray-50 p-4 rounded-lg border border-dashed border-gray-300">
            {/* Tabs */}
            <div className="flex space-x-4 mb-4 border-b pb-2">
              <button
                type="button"
                onClick={() => setActiveTab('upload')}
                className={`text-sm font-medium ${
                  activeTab === 'upload' ? 'text-blue-600 border-b-2 border-blue-600' : 'text-gray-500'
                }`}
              >
                Upload File
              </button>
              <button
                type="button"
                onClick={() => setActiveTab('link')}
                className={`text-sm font-medium ${
                  activeTab === 'link' ? 'text-blue-600 border-b-2 border-blue-600' : 'text-gray-500'
                }`}
              >
                External Link / YouTube
              </button>
            </div>

            {/* Upload Area */}
            {activeTab === 'upload' && (
              <div className="flex gap-2 flex-wrap">
                <label className="cursor-pointer flex items-center gap-2 px-4 py-2 bg-white border rounded hover:bg-gray-100 text-sm">
                  <ImageIcon size={16} /> Add Image
                  <input
                    type="file"
                    hidden
                    accept="image/*"
                    onChange={(e) => handleFileUpload(e)}
                  />
                </label>
                <label className="cursor-pointer flex items-center gap-2 px-4 py-2 bg-white border rounded hover:bg-gray-100 text-sm">
                  <Video size={16} /> Add Video
                  <input
                    type="file"
                    hidden
                    accept="video/*"
                    onChange={(e) => handleFileUpload(e)}
                  />
                </label>
                <label className="cursor-pointer flex items-center gap-2 px-4 py-2 bg-white border rounded hover:bg-gray-100 text-sm">
                  <FileText size={16} /> Add PDF
                  <input
                    type="file"
                    hidden
                    accept=".pdf"
                    onChange={(e) => handleFileUpload(e)}
                  />
                </label>
              </div>
            )}

            {/* Link Area */}
            {activeTab === 'link' && (
              <div className="flex gap-2">
                <input
                  type="text"
                  placeholder="Paste URL here..."
                  className="flex-1 p-2 border rounded text-sm"
                  value={linkInput}
                  onChange={(e) => setLinkInput(e.target.value)}
                />
                <button
                  type="button"
                  onClick={() => handleAddLink('youtube')}
                  className="bg-red-600 text-white px-3 py-1 rounded text-sm flex items-center gap-1"
                >
                  <Youtube size={14} /> YouTube
                </button>
                <button
                  type="button"
                  onClick={() => handleAddLink('link')}
                  className="bg-blue-600 text-white px-3 py-1 rounded text-sm flex items-center gap-1"
                >
                  <LinkIcon size={14} /> Link
                </button>
              </div>
            )}
          </div>

          {/* Media Preview */}
          {mediaItems.length > 0 && (
            <div className="mt-4 grid grid-cols-2 md:grid-cols-4 gap-4">
              {mediaItems.map((item, index) => (
                <div
                  key={index}
                  className="relative group border rounded-lg overflow-hidden bg-gray-100 aspect-video flex items-center justify-center"
                >
                  {/* Preview */}
                  {item.type === 'image' && (
                    <img src={item.url} alt="preview" className="w-full h-full object-cover" />
                  )}
                  {item.type === 'youtube' && item.thumbnail && (
                    <img src={item.thumbnail} alt="YouTube" className="w-full h-full object-cover" />
                  )}
                  {item.type === 'pdf' && (
                    <div className="text-center p-2">
                      <FileText className="mx-auto mb-1 text-red-500" />
                      <span className="text-xs text-gray-600 block truncate w-full">PDF Document</span>
                    </div>
                  )}
                  {(item.type === 'link' || item.type === 'video') && (
                    <div className="text-center">
                      <LinkIcon className="mx-auto mb-1 text-blue-500" />
                      <span className="text-xs text-gray-600">External Resource</span>
                    </div>
                  )}

                  {/* Remove Button */}
                  <button
                    type="button"
                    onClick={() => removeMedia(index)}
                    className="absolute top-1 right-1 bg-red-500 text-white p-1 rounded-full opacity-0 group-hover:opacity-100 transition-opacity"
                  >
                    <Trash2 size={12} />
                  </button>

                  {/* Type Badge */}
                  <span className="absolute bottom-1 left-1 bg-black/50 text-white text-[10px] px-1 rounded uppercase">
                    {item.type}
                  </span>
                </div>
              ))}
            </div>
          )}
        </div>

        <button
          type="submit"
          disabled={loading}
          className="w-full bg-indigo-600 text-white py-3 rounded-lg font-semibold hover:bg-indigo-700 transition disabled:bg-gray-400"
        >
          {loading ? 'Processing...' : 'Submit Content'}
        </button>
      </form>
    </div>
  );
}
